c10/macros/cmake_macros.h.in

```diff
@@ -9,5 +9,6 @@
 #cmakedefine C10_USE_GFLAGS
 #cmakedefine C10_USE_NUMA
 #cmakedefine C10_USE_MSVC_STATIC_RUNTIME
+#cmakedefine TORCH_DISABLE_GPU_ASSERTS
```

Comment:
we were looking at where TORCH_DISABLE_GPU_ASSERTS is defined for use by c10/macros/Macros.h, and thought this will make it visible to it.

Reply:
In other words, without this, the kernel assert will be still enabled for ROCm:
```
root@hpe-sjc2-36:/dockerx/development/pytorch2/pytorch# python
Python 3.9.18 (main, Sep 11 2023, 13:41:44) 
[GCC 11.2.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> import sys
>>> r=subprocess.call([sys.executable, '-c', "import torch;torch._assert_async(torch.tensor(0,device='cuda'));torch.cuda.synchronize()"])
/dockerx/development/pytorch2/pytorch/aten/src/ATen/native/hip/TensorCompare.hip:108: _assert_async_cuda_kernel: Device-side assertion `input[0] != 0' failed.
:0:rocdevice.cpp            :2690: 1808994712659 us: [pid:877166 tid:0x7f8307f31700] Callback: Queue 0x7f8204800000 aborting with error : HSA_STATUS_ERROR_EXCEPTION: An HSAIL operation resulted in a hardware exception. code: 0x1016
>>> 
```

Reply:
> @hongxiayang Looks like TORCH_DISABLE_GPU_ASSERTS was already used in
> 
> ```
> c10/macros/Macros.h
> ```
> 
> before this PR, hence this define is probably not needed

yes, it was used there to do conditional check. However,  how to make the CMakeLists.txt's cmake option 
```option(TORCH_DISABLE_GPU_ASSERTS "Disable GPU asserts by default" OFF)
```
or https://github.com/pytorch/pytorch/blob/6ae0554d11b973930d7b8ec1e937b27ac961d7bf/cmake/Dependencies.cmake#L1202C8-L1202C8
translated to
either 
```
#define TORCH_DISABLE_GPU_ASSERTS 
```
or 
```
/* #undef TORCH_DISABLE_GPU_ASSERTS */ 
```

Reply:
@malfet Does this address your concerns?

Reply:
Ok, I see now. 
So  TORCH_DISABLE_GPU_ASSERTS was introduced by https://github.com/pytorch/pytorch/pull/84190 but wasn't really tested working (nor tested) and you are trying to fix it here. 

Please note that there are multiple way of propagating CMAKE options into build flags. One is by using `.h.in` files and another is by invoking `target_compile_definitions(` or (which is blunt as it applies to all libraries) by calling `string(APPEND CMAKE_CXX_FLAGS`

Benefit of one method over another is whether the flag is carried over to extensions or not.

I'm ok with having it in `.h.in` file, though it feels it should be part of  `c10/cuda/impl/cuda_cmake_macros.h.in` as it is specific to GPUs isn't it?

Reply:
Just verified that, simply put that `#cmakedefine` in `c10/cuda/impl/cuda_cmake_macros.h.in` did not work, as the `CUDA_KERNEL_ASSERT` is defined in `c10/macros/Macros.h`, and `c10/macros/Macros.h` does not depend on `c10/cuda/CUDAMacros.h` file. Therefore the #define line in c10/cuda/impl/cuda_cmake_macros.h (or c10/hip/impl/hip_cmake_macros.h) has no impact on `CUDA_KERNEL_ASSERT`.  

Unless, we move the whole logic of  `#define CUDA_KERNEL_ASSERT` definition too, which will be a much bigger change?

Reply:
or, let `c10/cuda/CUDAMacros.h` include `c10/macros/Macros.h`, and replace those places in kernel code to include `c10/cuda/CUDAMacros.h` instead of `c10/macros/Macros.h`. Still sounds like much bigger change.

Reply:
Sorry, I didn't realize that CUDA_KERNEL_ASSERT is part of macros then cuda_macros.h. In that case, PR looks good to me thank you for the explanaiton.